// @microsoft/mixed-reality-extension-sdk/built/types/runtime/collider.js
"use strict";
/*!
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
const collider_1 = require("../internal/collider");
/**
 * A collider represents the abstraction of a physics collider object on the host.
 */
class Collider {
    /**
     * @hidden
     * Creates a new Collider instance.
     * @param $owner The owning actor instance. Field name is prefixed with a dollar sign so that it is ignored by
     * the actor patch detection system.
     * @param initFrom The collider like to use to init from.
     */
    constructor($owner, from) {
        this.$owner = $owner;
        this.$DoNotObserve = ['_internal'];
        this.enabled = true;
        this.isTrigger = false;
        if (from) {
            if (!from.geometry || !from.geometry.shape) {
                throw new Error("Must provide valid collider params containing a valid shape");
            }
            this._internal = new collider_1.InternalCollider(this, $owner);
            if (from.geometry !== undefined)
                this.geometry = from.geometry;
            if (from.enabled !== undefined)
                this.enabled = from.enabled;
            if (from.isTrigger !== undefined)
                this.isTrigger = from.isTrigger;
        }
        else {
            throw new Error("Must provide a valid collider-like to initialize from.");
        }
    }
    // public collisionLayer = CollisionLayer.Object;
    /** @hidden */
    get internal() { return this._internal; }
    /**
     * The current event subscriptions that are active on this collider.
     */
    get eventSubscriptions() {
        return this.internal.eventSubscriptions;
    }
    /**
     * Add a collision event handler for the given collision event state.
     * @param eventType The type of the collision event.
     * @param handler The handler to call when a collision event with the matching
     * collision event state is received.
     */
    onCollision(eventType, handler) {
        this.internal.on(eventType, handler);
    }
    /**
     * Remove the collision handler for the given collision event state.
     * @param eventType The type of the collision event.
     * @param handler The handler to remove.
     */
    offCollision(eventType, handler) {
        this.internal.off(eventType, handler);
    }
    /**
     * Add a trigger event handler for the given collision event state.
     * @param eventType The type of the trigger event.
     * @param handler The handler to call when a trigger event with the matching
     * collision event state is received.
     */
    onTrigger(eventType, handler) {
        this.internal.on(eventType, handler);
    }
    /**
     * Remove the trigger handler for the given collision event state.
     * @param eventType The type of the trigger event.
     * @param handler The handler to remove.
     */
    offTrigger(eventType, handler) {
        this.internal.off(eventType, handler);
    }
    /** @hidden */
    toJSON() {
        return {
            enabled: this.enabled,
            isTrigger: this.isTrigger,
            // collisionLayer: this.collisionLayer,
            geometry: this.geometry,
            eventSubscriptions: this.eventSubscriptions
        };
    }
}
exports.Collider = Collider;
